File: frontend/src/analytics/registry.js
// E09-02: client-side mirror of the backend analytics event registry
// (backend/app/analytics/registry.py). Keeps malformed payloads from ever
// leaving the device; the backend registry remains the authoritative
// contract. Extend both together (plus the CHECK constraints in
// backend/migrations/analytics_events.sql) when new events are approved.
//
// Privacy envelope (owner decision D1): events are strictly anonymous.
// Never declare properties that could carry user IDs, resource IDs, URLs,
// coordinates, or free text -- only closed enums of coarse values.

export const SCREEN_NAMES = Object.freeze([
  'map',
  'game_details',
  'profile',
  'notifications',
  'admin',
])

// Seed events approved in owner decision D2.
const EVENT_REGISTRY = Object.freeze({
  app_open: Object.freeze({}),
  screen_view: Object.freeze({
    screen: Object.freeze({ allowedValues: SCREEN_NAMES, required: true }),
  }),
})

export function isRegisteredEvent(eventName) {
  return Object.prototype.hasOwnProperty.call(EVENT_REGISTRY, eventName)
}

export function validateEvent(eventName, properties = {}) {
  if (!isRegisteredEvent(eventName)) {
    return [`unknown event_name: ${eventName}`]
  }
  if (!properties || typeof properties !== 'object' || Array.isArray(properties)) {
    return ['properties must be a plain object']
  }

  const errors = []
  const spec = EVENT_REGISTRY[eventName]

  for (const propertyName of Object.keys(properties)) {
    if (!Object.prototype.hasOwnProperty.call(spec, propertyName)) {
      errors.push(`unknown property ${propertyName} for event ${eventName}`)
    }
  }

  for (const [propertyName, propertySpec] of Object.entries(spec)) {
    if (!Object.prototype.hasOwnProperty.call(properties, propertyName)) {
      if (propertySpec.required) {
        errors.push(`missing required property ${propertyName} for event ${eventName}`)
      }
      continue
    }

    const value = properties[propertyName]
    if (typeof value !== 'string') {
      errors.push(`property ${propertyName} for event ${eventName} must be a string`)
      continue
    }
    if (!propertySpec.allowedValues.includes(value)) {
      errors.push(`invalid value for property ${propertyName} of event ${eventName}`)
    }
  }

  return errors
}

export function isValidEvent(eventName, properties = {}) {
  return validateEvent(eventName, properties).length === 0
}

// App-level route -> approved screen name. Routes without an approved enum
// value (my-games, my-reports) are intentionally not tracked; the
// 'game_details' and 'notifications' enum values are reserved for later
// modal-level instrumentation (wiring them now would require touching
// MapPage/NotificationsModal, which E09-02 must not modify).
export function screenNameForPathname(pathname) {
  if (pathname === '/') {
    return 'map'
  }
  if (pathname === '/admin') {
    return 'admin'
  }
  if (pathname === '/settings') {
    return 'profile'
  }
  return null
}

File: backend/app/analytics/registry.py
"""E09-02: declarative registry for first-party analytics events.

The registry is the single contract for what the ingestion pipeline accepts.
Every event name and every property must be declared here with a bounded set
of allowed values; anything else is rejected before it can reach the
database. Future issues add events by extending EVENT_REGISTRY (and the
matching CHECK constraints in backend/migrations/analytics_events.sql plus
the client mirror in frontend/src/analytics/registry.js).

Privacy envelope (owner decision D1): events are strictly anonymous. The
registry must never declare properties that could carry user IDs, resource
IDs, URLs, coordinates, or free text -- only closed enums of coarse values.
"""

from dataclasses import dataclass, field
from typing import Any, Mapping

SCREEN_NAMES: frozenset[str] = frozenset(
    {
        "map",
        "game_details",
        "profile",
        "notifications",
        "admin",
    }
)


@dataclass(frozen=True)
class PropertySpec:
    """A single allowed event property with a closed set of values."""

    allowed_values: frozenset[str] = field(default_factory=frozenset)
    required: bool = True


# Seed events approved in owner decision D2. Property values must always be
# closed enums -- never free text.
EVENT_REGISTRY: dict[str, dict[str, PropertySpec]] = {
    "app_open": {},
    "screen_view": {
        "screen": PropertySpec(allowed_values=SCREEN_NAMES, required=True),
    },
}


def is_registered_event(event_name: str) -> bool:
    return event_name in EVENT_REGISTRY


def validate_event(event_name: str, properties: Mapping[str, Any]) -> list[str]:
    """Validate an event against the registry.

    Returns a list of human-readable validation errors; an empty list means
    the event conforms to the registry contract.
    """
    if not is_registered_event(event_name):
        return [f"unknown event_name: {event_name!r}"]

    errors: list[str] = []
    spec = EVENT_REGISTRY[event_name]

    for property_name in properties:
        if property_name not in spec:
            errors.append(
                f"unknown property {property_name!r} for event {event_name!r}"
            )

    for property_name, property_spec in spec.items():
        if property_name not in properties:
            if property_spec.required:
                errors.append(
                    f"missing required property {property_name!r} for event {event_name!r}"
                )
            continue

        value = properties[property_name]
        if not isinstance(value, str):
            errors.append(
                f"property {property_name!r} for event {event_name!r} must be a string"
            )
            continue
        if value not in property_spec.allowed_values:
            errors.append(
                f"invalid value for property {property_name!r} of event {event_name!r}"
            )

    return errors
